### ISSUE TITLE
Nested effects.

### ISSUE THREAD

**Issue by kethan** [preactjs/signals#114]
I have an issue if I nest effects like this:
```js
const counter = signal(1);
const double = computed(() => counter.value * 2);
const tripple = computed(() => counter.value * 3);

effect(() => {
  console.log(double.value);
  effect(() => {
    console.log(tripple.value);
  });
});
counter.value = 20;
```
The triple value is consoled 3 times.

**Comment by marvinhagemeister** [comment]
Curious to learn more. What's a common use case for nested effects?

**Comment by kethan** [comment]
Hmmm. not much just thinking of solidjs cleaning up of effects. If I am making a general JSX library components can be nested with effects so I want to clean up those instead of piling up.

**Comment by marvinhagemeister** [comment]
That's awesome. For now there are no plans to support nested effects. If you're looking for something solid-like the best options is probably to just use solid itself.

**Comment by fabiospampinato** [comment]
> Curious to learn more. What's a common use case for nested effects?

I like this use case: I have a bunch of effects that need to be active if and only if the component is "open", you can just write a parent effect that's re-executed when the open state changes, and the nested effects can exist potentially only when the component is open, just naturally like that, like putting them inside a reactive `if` statement.

Basically nesting can potentially turn N checks for N effects into 1 check and N+1 effects.

**Comment by kethan** [comment]
Eg: https://codesandbox.io/s/signal-clean-effect-torrpe?file=/src/index.js

**Comment by marvinhagemeister** [comment]
Fixed by #158

**Comment by marvinhagemeister** [comment]
Little update: We decided to revert #158 . Was a bit of a discussion but ultimately, but in the end we all felt that supporting auto-disposal of nested effects is outside of the scope for this project. With #183 the disposals of inner effects can be done by collected them in something like an array and passing them to the outer cleanup function.

```jsx
effect(() => {
  const dispose = effect(() => {
    // ...
  });

  // The cleanup function is run whenever the effect is disposed or re-run again
  return () => {
    dispose();
  };
});
```

This is way nested effects are fully supported, but we don't want to encourage that.

We started signals with the main goal of making working with state awesome in Preact and then branch out to other frameworks in the form of adapters. The use cases which have been brought up here so far relate to making signals the framework itself. That space is already well covered by solidjs. There is nothing wrong with just using solidjs directly. The other scenarios that we've seen so far, are scenarios which become much more readable once re-written to not rely on nesting effects. This lead us to conclude that we don't want to encourage users to rely on nested effects in their apps.
